### ISSUE TITLE
App crash on large chunk of data on Android <= 12

### ISSUE THREAD

**Issue by hkellou-vitalcare** [Abhinandan-Kushwaha/react-native-gifted-charts#437]
**Describe the bug**
Hello there, I've encountered a bug where large chunks of data (250+) displayed in LineChart will crash the app instantly on Android devices lower than android 13. I've encountered the error in Android 12, 11, and 10. The app also doesn't show the linechart on Android 7, it stays empty. It does work as intended on Android 13 tho. (didn't try on other devices)

I have tried different versions of the package, 
- 1.3.16 
- 1.3.12 (which i was using originally)
- 1.3.2
- 1.2.42

**To Reproduce**
1. Send large chunks of data (over 250)
2. Open the app
3. Instant crash

I was able to reproduce the bug on a Snack project under Expo, so you can try it by yourself.
https://snack.expo.dev/@doctibbers/rn-gifted-charts-large-amount-bug?platform=android

It will work fine on any Android 13+ device and crashes on the rest.

**Expected behavior**
App should just show the normal data like the Android 13 behaviour

**Screenshots**
![Screenshot_1700159526](https://github.com/Abhinandan-Kushwaha/react-native-gifted-charts/assets/121389649/047e592f-190d-4c20-abdd-745afebdec7c)

 - Device: Any android device
 - OS: Android 12, 11, 10 (other versions also need to be tried)
 - Version of the library being used: 1.3.16, 1.3.12, 1.3.2, 1.2.42

**Comment by hkellou-vitalcare** [comment]
@Abhinandan-Kushwaha Hi there. I haven't heard from you back yet. Were you able to reproduce this issue ? I might provide more information if needed. 
I am also trying to figure out what's going on from my side. Please let me know if you need anything.
Thank you for your work!

**Comment by Abhinandan-Kushwaha** [comment]
Hi @hkellou-vitalcare 
I tried but was unable to repro the issue on Android 11.
I rendered charts with datasets bigger than 300. (On Simulator)

**Comment by govind-appnox** [comment]
Hello Guys I'm facing the same issues Please can someone help

**Comment by hkellou-vitalcare** [comment]
> Hi @hkellou-vitalcare I tried but was unable to repro the issue on Android 11. I rendered charts with datasets bigger than 300. (On Simulator)

I was also able to run it on a Samsung A70 under Android 11 with no issues. However on Honor 8, Cat 67, and on simulator it crashes.
I do believe it might be related to the amount of RAM available or something. It's like a safe shutdown of the app

Try with my snack reproduction here and a simulator on android 11 with no more than 2gb of ram, it might be this schema that will crash for you
https://snack.expo.dev/@doctibbers/rn-gifted-charts-large-amount-bug?platform=android

**Comment by hkellou-vitalcare** [comment]
> Hello Guys I'm facing the same issues Please can someone help

For now, i'm unable to find a fix yet. I have limited the number of measures / data i show to something that doesn't crash meanwhile i investigate for a long-term solution.
You might want to do something like that to avoid crashes meanwhile

**Comment by hkellou-vitalcare** [comment]
@Abhinandan-Kushwaha I have a crazy update for you. 
We bought literally 3 models of the same phone (Samsung M13) with 4GO of ram, same characteristics, same everything.
It works fine on one, and crashes on 2 devices. Same account, same data.
I have no idea what's going on at this point. Large chunks of data behave differently and inconsistently.

**Comment by hkellou-vitalcare** [comment]
@Abhinandan-Kushwaha Dear owner, i'm back with a stack trace of the logcat i get on app crash
`11-29 12:21:47.561 19617 19617 E AndroidRuntime: FATAL EXCEPTION: main
11-29 12:21:47.561 19617 19617 E AndroidRuntime: Process: com.app, PID: 19617
11-29 12:21:47.561 19617 19617 E AndroidRuntime: java.lang.RuntimeException: Canvas: trying to draw too large(105676224bytes) bitmap.
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.graphics.RecordingCanvas.throwIfCannotDraw(RecordingCanvas.java:280)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.graphics.BaseRecordingCanvas.drawBitmap(BaseRecordingCanvas.java:69)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.horcrux.svg.SvgView.onDraw(SvgView.java:133)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.draw(View.java:22350)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21226)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.draw(View.java:22081)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.drawChild(ViewGroup.java:4516)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.facebook.react.views.view.ReactViewGroup.drawChild(ReactViewGroup.java:836)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4277)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.facebook.react.views.view.ReactViewGroup.dispatchDraw(ReactViewGroup.java:809)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21217)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.draw(View.java:22081)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.drawChild(ViewGroup.java:4516)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.facebook.react.views.view.ReactViewGroup.drawChild(ReactViewGroup.java:836)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4277)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.facebook.react.views.view.ReactViewGroup.dispatchDraw(ReactViewGroup.java:809)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21217)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.recreateChildDisplayList(ViewGroup.java:4500)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewGroup.dispatchGetDisplayList(ViewGroup.java:4473)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.View.updateDisplayListIfDirty(View.java:21186)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:559)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:565)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:642)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewRootImpl.draw(ViewRootImpl.java:4101)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:3828)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3099)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1952)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8171)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.Choreographer$CallbackRecord.run(Choreographer.java:972)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.Choreographer.doCallbacks(Choreographer.java:796)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.Choreographer.doFrame(Choreographer.java:731)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:957)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7656)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
11-29 12:21:47.561 19617 19617 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)`

Basically the error is that `java.lang.RuntimeException: Canvas: trying to draw too large(105676224bytes) bitmap.`
Maybe this would help figure out the origin of the problem

**Comment by Abhinandan-Kushwaha** [comment]
Hi @hkellou-vitalcare I think Android does not aloow rendering very large svgs. See https://github.com/software-mansion/react-native-svg/issues/440

We should probably implement pagination (scroll to load) for very large scrollable charts. I will give it a try in some time.

**Comment by hkellou-vitalcare** [comment]
Hi @Abhinandan-Kushwaha. Thanks for your answer! That's actually what i was thinking in the first place, that the RAM might be unsufficient in those phones. 
I would suggest to add something like the "onEndReached"  prop used on FlatList, so that anyone can be able to introduce some kind of "infinite" pagination/scrolling. I can see you have a "scrollToEnd" prop, if that's already using some function to calculate the end of the graphic, it can be used to introduce a sort of pagination using something like RTK Query.

**Comment by hkellou-vitalcare** [comment]
Hi @Abhinandan-Kushwaha 
Are you planning on introducing anything that would permit to not load the whole data at once to fix this particular issue?
I know it's a bit tricky and needs some workaround but wanted to know if you had any timeline to implement a fix/workaround
Thank you for you work

**Comment by hkellou-vitalcare** [comment]
@Abhinandan-Kushwaha Thank you for your hard work!

**Comment by AlejandroRM-DEV** [comment]
I'm facing the same situation, I'll wait this feature, thanks @Abhinandan-Kushwaha

**Comment by hkellou-vitalcare** [comment]
I have opened a PR trying to implement the method needed to achieve pagination on data, please feel free to review it @Abhinandan-Kushwaha

**Comment by Abhinandan-Kushwaha** [comment]
@hkellou-vitalcare Thanks for this much-needed PR. I have merged it and will release a new version soon.

**Comment by Abhinandan-Kushwaha** [comment]
Hi @hkellou-vitalcare @govind-appnox 
It is now available from version `1.4.1` onwards.
I have renamed the props to- `onStartReached` and `onEndReached`.
Also added an extra prop named `endReachedOffset`.

See [pagination related props](https://github.com/Abhinandan-Kushwaha/react-native-gifted-charts/blob/master/docs/BarChart/BarChartProps.md#pagination-related-props)

**Comment by ftaibi** [comment]
I realised that if areaChart or curved are enabled the performance is affected a lot the JS thread fps goes down
